libbeat/registry/backend/cptest/util.go

```diff
@@ -0,0 +1,122 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cptest
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+)
+
+// RunWithPath uses the factory to create and configure a registry with a
+// temporary test path. The test function fn is called with the new reigstry.
+// The registry is closed once the test finishes and the temporary is deleted
+// afterwards (unless the `-keep` CLI flag is used).
+func RunWithPath(t *testing.T, factory BackendFactory, fn func(*Registry)) {
+	WithPath(factory, func(_ *testing.T, reg *Registry) {
+		fn(reg)
+	})(t)
+}
+
+// WithPath wraps a registry aware test function into a normalized test
+// function that can be used with `t.Run`.
+// The factory is used to create and configure the reigstry with a temporary
+// test path.  The registry is closed and the temporary test directoy is delete
+// if the test function returns or panics.
+func WithPath(factory BackendFactory, fn func(*testing.T, *Registry)) func(t *testing.T) {
+	return func(t *testing.T) {
+		path, err := ioutil.TempDir(defaultTempDir, "")
+		if err != nil {
+			t.Fatalf("Failed to create temporary test directory: %v", err)
+		}
+
+		t.Logf("Test tmp dir: %v", path)
+		if !keepTmpDir {
+			defer os.RemoveAll(path)
+		}
+
+		reg, err := factory(path)
+		if err != nil {
+			t.Fatalf("Failed to create registry: %v", err)
+		}
+		defer reg.Close()
+
+		fn(t, &Registry{T: t, Registry: reg})
+	}
+}
```

Comment:
nit: for bookkeeping of temporary files should we recover from a panic to still remove the temporary files?

Reply:
This functions uses `defer os.RemoveAll(path)`. With `defer` there is no need to recover from panics, as deferred functions are always run, even in the presence of recoverable panics.
There is a second kind of panic, which you can not recover from. Plain old segmentation faults. If these occur (very unlikely), the process just crashes very hard (still with a panic).

For debugging an issue it can be helpful to not delete the test directories. This is the reason this package adds the CLI flags `-dir` and `-keep` (I used them to create sample stores in the next PR).